internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,497 @@
+const (
+	JOB_OUTPUT_LOCK_FORMAT     = "%s::%s"
+	JOB_OUTPUT_ERROR           = "error"
+	JOB_OUTPUT_WORK_ERRORS     = "work-errors"
+	JOB_OUTPUT_RESULT_ITEMS    = "result-items"
+	JOB_OUTPUT_REMAINING_ITEMS = "remaining-items"
+	JOB_OUTPUT_RUN_TIME        = "run-time"
+)
+
+// JobOptions contains user level options for a Job.
+// These options allow for optional context cancellation of the Job.
+type JobOptions struct {
+	// Don't save anything.
+	DryRun bool `json:"dry-run"`
+
+	// Remove any existing records before running the job.
+	OverwriteRecords bool `json:"overwrite-records"`
+
+	// Wait for the entire job to complete and return all results.
+	WaitForCompletion bool `json:"wait-for-completion"`
+
+	// Return a copy of an intermediate result.
+	// Will not be able to access final output.
+	// If WaitForCompletion is true, this option is overwritten.
+	ReturnCopy bool `json:"return-copy"`
+
+	// A context that can be used to cancel the job.
+	Context context.Context `json:"-"`
+}
+
+// Job provides system level customization of the job's execution.
+// It supports the following optional functionality:
+//   - synchronization
+//   - context cancellation
+//   - record retrieval
+//   - single record retrieval
+//   - record storage
+//   - record removal from storage
+//   - stat tracking
+//
+// Given job options, input items, and a work function,
+// Job will process a list of input items through the work func in a parallel pool to produce a JobOutput.
+// If a lock key is provided, Job will block on that key before running preventing overuse of resources or conflicts between the same Job.
+// Provide an input key generation function to synchronize at the input item level.
+type Job[InputType any, OutputType any] struct {
+	// The user level options for a Job.
+	*JobOptions
+
+	// The number of workers in the parallel pool.
+	PoolSize int
+
+	// An optional key to lock on.
+	LockKey string
+
+	// A sorted list of items that need work.
+	WorkItems []InputType
+
+	// An optional function to retrieve existing records that should not be processed.
+	// Returns a list of processed records, remaining items, and an error.
+	RetrieveFunc func([]InputType) ([]OutputType, []InputType, error)
+
+	// An optional function to retrieve a record based on a single work item.
+	SingleRetreiveFunc func(InputType) (OutputType, bool, error)
+
+	// An optional function to store the result.
+	StoreFunc func([]OutputType) error
+
+	// An optional function to remove existing records from storage.
+	RemoveStorageFunc func([]InputType) error
+
+	// A function to transform work items into results.
+	// Returns a result, the time of computation, and an error.
+	WorkFunc func(InputType) (OutputType, error)
+
+	// An optional function to get the locking key for an individual work item.
+	WorkItemKeyFunc func(InputType) string
+
+	// An optional function to store stats on how long the job took.
+	StatFunc func(int64)
```

Comment:
I don't see a strong reason to make this specific to stats.
This could just be some function that gets called with the full JobOutput object when the job is done.
Maybe "OnComplete()"?